File: incre-tests/synduce/combine/mts.f
Inductive List = nil Unit | cons {Int, List};

max = \a: Int. \b: Int. if < a b then b else a;

sum = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> + h (f t)
  end
);

mts = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> max (+ h (sum t)) (f t)
  end
);

spec = \xs: List. mts xs;

repr = fix (
  \f: List -> Compress List. \xs: List.
  match xs with
    nil _ -> nil unit
  | cons {h, t} -> cons {h, f t}
  end
);

main = \xs: List. spec (repr xs);

File: ml-benchmark/synduce/combine/mts.ml
type 'a list =
  | Nil
  | Cons of 'a * 'a list

let rec sum = function
  | Nil -> 0
  | Cons (hd, tl) -> hd + sum tl

let max a b = if a > b then a else b

let rec mts = function
  | Nil -> 0
  | Cons (hd, tl) -> max (hd + sum tl) (mts tl)

let spec l = (sum l, mts l)

val repr: int list -> (int list) compress
let rec repr = function 
  | Nil -> Nil 
  | Cons (h, t) -> Cons (h, repr t)

let prog xs = spec (repr xs)